include/unifex/upon_done.hpp

```diff
@@ -0,0 +1,169 @@
+#pragma once
+
+#include <unifex/config.hpp>
+#include <unifex/async_trace.hpp>
+#include <unifex/bind_back.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/std_concepts.hpp>
+#include <unifex/tag_invoke.hpp>
+#include <unifex/type_list.hpp>
+#include <unifex/type_traits.hpp>
+
+#include <functional>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+namespace _upon_done {
+
+template <typename Receiver, std::invocable Func>
+struct _receiver {
+  struct type;
+};
+template <typename Receiver, std::invocable Func>
+using receiver_t = typename _receiver<Receiver, Func>::type;
+
+template <typename Receiver, std::invocable Func>
+struct _receiver<Receiver, Func>::type {
+  UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+  UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+
+  template <typename... Values>
+  void set_value(Values&&... values) && noexcept {
+    unifex::set_value(std::move(receiver_), std::forward<Values>(values)...);
+  }
+
+  template <typename Error>
+  void set_error(Error&& error) && noexcept {
+    unifex::set_error(std::move(receiver_), std::forward<Error>(error));
+  }
+
+  void set_done() && noexcept {
+    if constexpr (noexcept(std::invoke(std::move(func_)))) {
+      std::invoke(std::move(func_));
+      unifex::set_done(std::move(receiver_));
+    } else {
+      UNIFEX_TRY {
+        std::invoke(std::move(func_));
+        unifex::set_value(std::move(receiver_));
+      }
+      UNIFEX_CATCH(...) {
+        unifex::set_error(std::move(receiver_), std::current_exception());
+      }
+    }
+  }
+
+  template <typename CPO, typename R>
+  requires is_receiver_query_cpo_v<CPO> AND same_as<R, type>
```

Comment:
We use a `template` macro (I know) to make it possible to use `requires` in C++17 code. It maps to `enable_if` under the covers.

```suggestion
  template (typename CPO, typename R)
    (requires is_receiver_query_cpo_v<CPO> AND same_as<R, type>)
```

And similar changes elsewhere.